mirgecom/simutil.py

```diff
@@ -290,7 +299,36 @@ def compare_fluid_solutions(discr, red_state, blue_state):
     """
     actx = red_state.array_context
     resid = red_state - blue_state
-    return [actx.to_numpy(discr.norm(v, np.inf)) for v in resid.join()]
+    resid_errs = actx.to_numpy(
+        flatten(componentwise_norms(discr, resid, order=np.inf), actx))
+
+    return resid_errs.tolist()
+
+
+def componentwise_norms(discr, fields, order=np.inf):
+    """Return the *order*-norm for each component of *fields*.
+
+    .. note::
+        This is a collective routine and must be called by all MPI ranks.
+    """
+    if not isinstance(fields, DOFArray):
+        return map_array_container(
+            partial(componentwise_norms, discr, order=order), fields)
+    if len(fields) > 0:
+        return discr.norm(fields, order)
+    else:
+        return 0
```

Comment:
Ah this is for handling cases where `species_mass` is an empty object array. At the moment, grudge reductions do not handle being passed an empty object array in a container (it messes up quite a lot surprisingly and I haven't been able to come up with a clean workaround). So this is an attempt to filter that.

I think one could argue that this type of special-casing should be done at the mirgecom level (since mirgecom is the one forcing down an empty object array in the first place). I would like to try to keep this special-casing out of grudge if possible.

Reply:
@majosm @thomasgibson we have a "better" fix for this that is coming in an upcoming PR.  Namely, the SplitCV strategy where we'll just use a version of a CV that does not carry around these dumb empty arrays for species components when there are no species.   

The current snippet here is just a work-around until that PR lands.

Reply:
fwiw: issue/pr #575 / #569